### ISSUE TITLE
Strange JVM crash with WebSocket Client

### ISSUE THREAD

**Issue by stoeckley** [stalefruits/gniazdo#29]
I've been trying for a couple days to track down why the entire JVM is crashing when I use this library sometimes (but not all times). I get error logs that start like this:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (sharedRuntime.cpp:834), pid=4715, tid=4103
#  fatal error: exception happened outside interpreter, nmethods and vtable stubs at pc 0x000000011e6b736f
#
# JRE version: Java(TM) SE Runtime Environment (8.0_45-b14) (build 1.8.0_45-b14)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.45-b02 mixed mode bsd-amd64 compressed oops)
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#

---------------  T H R E A D  ---------------

Current thread (0x00007f887cce4800):  JavaThread "WebSocketClient@931427653-45" [_thread_in_Java, id=4103, stack(0x000070000fcd4000,0x000070000fdd4000)]

Stack: [0x000070000fcd4000,0x000070000fdd4000],  sp=0x000070000fdd2130,  free space=1016k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.dylib+0x58e7ba]  VMError::report_and_die()+0x3f8
V  [libjvm.dylib+0x1e037f]  report_vm_error(char const*, int, char const*, char const*)+0x54
V  [libjvm.dylib+0x4c923f]  SharedRuntime::continuation_for_implicit_exception(JavaThread*, unsigned char*, SharedRuntime::ImplicitExceptionKind)+0x195
V  [libjvm.dylib+0x46fd57]  JVM_handle_bsd_signal+0x368
V  [libjvm.dylib+0x46c137]  signalHandler(int, __siginfo*, void*)+0x2f
C  [libsystem_platform.dylib+0x2b3a]  _sigtramp+0x1a
C  0x0000000136c7fd80
J 6250 C2 com.sun.crypto.provider.GaloisCounterMode.decryptFinal([BII[BI)I (256 bytes) @ 0x000000011f4c968f [0x000000011f4c8c00+0xa8f]


---------------  P R O C E S S  ---------------

Java Threads: ( => current thread )
  0x00007f887db85800 JavaThread "clojure-agent-send-pool-3" daemon [_thread_blocked, id=4363, stack(0x000070000cbb8000,0x000070000ccb8000)]
  0x00007f887def1000 JavaThread "clojure-agent-send-pool-2" daemon [_thread_blocked, id=71951, stack(0x00007000104ec000,0x00007000105ec000)]
  0x00007f8879154800 JavaThread "async-dispatch-8" daemon [_thread_blocked, id=72451, stack(0x00007000105ef000,0x00007000106ef000)]
  0x00007f887ea00800 JavaThread "async-dispatch-7" daemon [_thread_blocked, id=71427, stack(0x00007000103e9000,0x00007000104e9000)]
  0x00007f887ed69800 JavaThread "async-dispatch-6" daemon [_thread_blocked, id=70915, stack(0x00007000102e6000,0x00007000103e6000)]
 0x00007f887c93f800 JavaThread "clojure-agent-send-pool-1" daemon [_thread_blocked, id=70403, stack(0x00007000101e3000,0x00007000102e3000)]
  0x00007f887d82b000 JavaThread "WebSocketClient@931427653-scheduler" [_thread_blocked, id=69891, stack(0x00007000100e0000,0x00007000101e0000)]
  0x00007f8879384000 JavaThread "WebSocketClient@931427653-48" [_thread_in_native, id=69379, stack(0x000070000ffdd000,0x00007000100dd000)]
  0x00007f8879383000 JavaThread "WebSocketClient@931427653-47" [_thread_blocked, id=41223, stack(0x000070000feda000,0x000070000ffda000)]
  0x00007f887e94a800 JavaThread "WebSocketClient@931427653-46" [_thread_blocked, id=40975, stack(0x000070000fdd7000,0x000070000fed7000)]
=>0x00007f887cce4800 JavaThread "WebSocketClient@931427653-45" [_thread_in_Java, id=4103, stack(0x000070000fcd4000,0x000070000fdd4000)]
  0x00007f887cce4000 JavaThread "WebSocketClient@931427653-44" [_thread_in_native, id=47627, stack(0x000070000fbd1000,0x000070000fcd1000)]
  0x00007f887f10e000 JavaThread "WebSocketClient@931427653-43" [_thread_in_native, id=62483, stack(0x000070000face000,0x000070000fbce000)]
  0x00007f887e969800 JavaThread "WebSocketClient@931427653-42" [_thread_in_native, id=60963, stack(0x000070000f9cb000,0x000070000facb000)]
  0x00007f88794ac000 JavaThread "WebSocketClient@931427653-41" [_thread_blocked, id=62727, stack(0x000070000e280000,0x000070000e380000)]
  0x00007f887ed13800 JavaThread "async-dispatch-5" daemon [_thread_in_Java, id=68615, stack(0x000070000f8c8000,0x000070000f9c8000)]
  0x00007f887e8c9800 JavaThread "clojure-agent-send-pool-0" daemon [_thread_blocked, id=68099, stack(0x000070000f7c5000,0x000070000f8c5000)]
  0x00007f887b078000 JavaThread "XNIO-1 Accept" daemon [_thread_in_native, id=67587, stack(0x000070000f6c2000,0x000070000f7c2000)]
  0x00007f887b077000 JavaThread "XNIO-1 I/O-8" daemon [_thread_in_native, id=67075, stack(0x000070000f5bf000,0x000070000f6bf000)]
  0x00007f887deeb000 JavaThread "XNIO-1 I/O-7" daemon [_thread_in_native, id=66563, stack(0x000070000f4bc000,0x000070000f5bc000)]
  0x00007f887e86b000 JavaThread "XNIO-1 I/O-6" daemon [_thread_in_native, id=66051, stack(0x000070000f3b9000,0x000070000f4b9000)]
  0x00007f887971d000 JavaThread "XNIO-1 I/O-5" daemon [_thread_in_native, id=65539, stack(0x000070000f2b6000,0x000070000f3b6000)]
  0x00007f887e8af800 JavaThread "XNIO-1 I/O-4" daemon [_thread_in_native, id=65027, stack(0x000070000f1b3000,0x000070000f2b3000)]
  0x00007f887e818800 JavaThread "XNIO-1 I/O-3" daemon [_thread_in_native, id=64515, stack(0x000070000f0b0000,0x000070000f1b0000)]
  0x00007f887924b800 JavaThread "XNIO-1 I/O-2" daemon [_thread_in_native, id=64003, stack(0x000070000efad000,0x000070000f0ad000)]
  0x00007f887ecbd000 JavaThread "XNIO-1 I/O-1" daemon [_thread_in_native, id=63491, stack(0x000070000eeaa000,0x000070000efaa000)]
  0x00007f8878c11800 JavaThread "async-dispatch-4" daemon [_thread_blocked, id=5639, stack(0x000070000eda7000,0x000070000eea7000)]
  0x00007f88787ba800 JavaThread "async-dispatch-3" daemon [_thread_blocked, id=45083, stack(0x000070000eca4000,0x000070000eda4000)]
  0x00007f887e080000 JavaThread "async-dispatch-2" daemon [_thread_blocked, id=28939, stack(0x000070000eba1000,0x000070000eca1000)]
  0x00007f887e32f800 JavaThread "async-dispatch-1" daemon [_thread_blocked, id=61195, stack(0x000070000ea9e000,0x000070000eb9e000)]
  0x00007f887f05c800 JavaThread "clojure-agent-send-off-pool-4" [_thread_in_native, id=59395, stack(0x000070000e795000,0x000070000e895000)]
  0x00007f887f10b800 JavaThread "clojure-agent-send-off-pool-3" [_thread_blocked, id=58883, stack(0x000070000e692000,0x000070000e792000)]
  0x00007f887a8e3000 JavaThread "clojure-agent-send-off-pool-2" [_thread_in_native, id=58371, stack(0x000070000e58f000,0x000070000e68f000)]
  0x00007f887dbe0000 JavaThread "AWT-AppKit" daemon [_thread_in_native, id=775, stack(0x00007fff506ee180,0x00007fff50f3f000)]
  0x00007f8879e3c800 JavaThread "clojure.core.async.timers/timeout-daemon" daemon [_thread_blocked, id=22275, stack(0x000070000dff7000,0x000070000e0f7000)]
  0x00007f8879812800 JavaThread "Service Thread" daemon [_thread_blocked, id=21251, stack(0x000070000ddf1000,0x000070000def1000)]
  0x00007f8878031800 JavaThread "C1 CompilerThread3" daemon [_thread_blocked, id=20739, stack(0x000070000dcee000,0x000070000ddee000)]
  0x00007f8879826800 JavaThread "C2 CompilerThread2" daemon [_thread_blocked, id=20227, stack(0x000070000dbeb000,0x000070000dceb000)]
```

I thought I'd post this here in case you've seen this before using gniazdo or can advise where I might look?

**Comment by mknoszlig** [comment]
Hi!
I haven't seen this bug before. From your error report it seems you're using a pretty old JRE: http://www.oracle.com/technetwork/java/javase/8u45-relnotes-2494160.html
Could you please try updating to a current version (1.8.0_181-b13 seems to most recent version of JDK8) and see if the crashes still happen?

cheers,

-Maxi

**Comment by stoeckley** [comment]
I updated the JRE to the latest and even restarted the machine for good measure, but the crash immediately occurred next time I received some WebSocket messages.

I did some further googling and found mention of this particular bug in a few places on the JIRA tracker for the JVM and it appears this is a "rare" bug that doesn't yet have a fix. It's not so rare for me, but nonetheless, the information suggested some JVM options to help mitigate it:

`"-XX:-TieredCompilation"`

and

`"-XX:-UseAESIntrinsics" `

I have been using the first one only for now and so far the crash appears to have been avoided. I have little understanding of JVM internals so it's not clear which of these would be the better to use, but I mention them here in case anyone else has this problem.
